### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by jt2002**
Pushed README, preprocess script, and model-building-saving script to the dev branch.  Next week, the script to generate recommendations.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,46 @@
...
+python build_save_model.py
+```
+
+Generate recommendation
+```
+python recnet.py
+```
+
+## Live Application
+[RecNet](https://www.wideanddeep.com/)
```
Comment (TensorDuck) [inline README.md:46]: This file is good! It's really good practice to always have a clean README at all times!

File: README.md [hunk 2]
```diff
@@ -1,2 +1,46 @@
...
+  - [Book (ratings only)](http://deepyeti.ucsd.edu/jianmo/amazon/categoryFilesSmall/Books.csv)
+  - [Book (metadata)](https://forms.gle/A8hBfPxKkKGFCP238)
+
+## Setup
+- Clone repository: `git clone https://github.com/jt2002/Insight.git`
+- Download data from [Amazon Review Data (2018)](https://nijianmo.github.io/amazon/index.html)
+- Rename `Books.csv` to `Book_review_rating.csv`
+- Move the CSV file to the folder `data\source_data`
+- Gunzip `meta_Books.json.gz` to `meta_Books.json`
+- Move the JSON file to the folder `data\source_data`
```
Comment (TensorDuck) [inline README.md:27]: To make this even better: Put all these steps in a shell script.

File: build_save_model.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    # Reset review_df to make this code section optional
+    review_df = scaled_df
+
+    # Reset cat.codes for tensorflow
+    review_df['userID'] = scaled_df['reviewerID'].astype('category').cat.codes
+    review_df['productID'] = scaled_df['asin'].astype('category').cat.codes
+
+    return review_df
+
+def load_review(product_pkl):
```
Comment (TensorDuck) [inline build_save_model.py]: I think you have duplicate function names here.

File: build_save_model.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+import pandas as pd
+from sklearn.utils import shuffle
+from datetime import datetime
+
+from tensorflow.keras.layers import Input, Embedding, Dense
+from tensorflow.keras.layers import Flatten, Concatenate
+from tensorflow.keras.models import Model
+from tensorflow.keras.optimizers import SGD
+
+def load_review(review_pkl):
```
Comment (TensorDuck) [inline build_save_model.py:25]: Duplicate function name

**Review by TensorDuck — COMMENTED** [review]
I pointed out duplicate function names. Not sure if it's accidental or not, but you should give them more distinctive names like `load_review_pickle" or `load_review_df`.

File: build_save_model.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    u_emb = Embedding(U, D)(u)        # (num_samples, 1, D)
+    p_emb = Embedding(P, D)(p)        # (num_samples, 1, D)
+    u_emb = Flatten()(u_emb)          # (num_samples, D)
+    p_emb = Flatten()(p_emb)          # (num_samples, D)
+    x = Concatenate()([u_emb, p_emb]) # (num_samples, 2D)
+    x = Dense(512, activation='relu')(x)
+    x = Dense(1)(x)
+    model = Model(inputs=(u,p), outputs=x)
+
+    return model
```
Comment (TensorDuck) [inline build_save_model.py:65]: This function is really well commented. To make it even better, I would say it needs a docstring in its final version.

File: build_save_model.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    print('Train-Test split\n')
+    Ntrain = int(train_test_split * len(rating))
+
+    train_user = user_id[:Ntrain]
+    train_product = product_id[:Ntrain]
+    train_rating = rating[:Ntrain]
+
+    test_user = user_id[Ntrain:]
+    test_product = product_id[Ntrain:]
+    test_rating = rating[Ntrain:]
```
Comment (TensorDuck) [inline build_save_model.py]: `sci-kit-learn` has a built-in train-test split that would probably suit your needs

File: build_save_model.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    r = model.fit(x=[train_user, train_product], y=train_rating,
+                  epochs=10,
+                  batch_size=1024,
+                  validation_data=([test_user, test_product], test_rating))
+    print('   Training end time = ', datetime.now().strftime("%H:%M:%S"), '\n')
+
+    print('... save model and input data')
+    model.save(saved_model)
+    review_cat_df.to_pickle(saved_input)
+    print(f'Done save model {saved_model} and input data {saved_input}')
```
Comment (TensorDuck) [inline build_save_model.py]: This file looks great! But it's hard to import and use (not modular). Great to have for now so keep it, but be ready to modularize before the end :thumbsup:

File: preprocess.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+                                dtype={'asin': object, 'reviewerID':object},
+                                sep=',')
+    except:
+        print(f'Error loading {review_file}\nExit!')
+        sys.exit(1)
+
+    print(f'  Done reading {review_file}')
+
+    review_df.drop_duplicates(subset=['reviewerID','asin'],
+                              keep='first', inplace=True)
```
Comment (TensorDuck) [inline preprocess.py]: this kind of formatting is totally fine! But you seem like you care about your code looking uniform and clean (which is really good). I just wanted to point out this tool that I think you might like: 
https://black.readthedocs.io/en/stable/

The typical use case would be to run the `black` linter on your whole directory before you push a commit. If you use VSCode, you can even set it to automatically trigger everytime you save a file.

File: preprocess.py [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+def preprocess_product(product_file,product_pkl):
+
+    chunksize = 100000
+    print(f'  Reading {product_file} in chunksize')
+    try:
+        chunk_df = pd.read_json(product_file, lines=True,
+                                dtype={'asin': object}, chunksize=chunksize)
+    except:
+        print(f'Error loading {product_file}\nExit!')
+        sys.exit(1)
```
Comment (TensorDuck) [inline preprocess.py:57]: This is not a bad pattern per-say. But your past-experience with functional programming languages is starting to show! More importantly, if this fails, it actually obscures the stack-trace so it doesn't make it clear why it failed. I would recommend removing the try-except block entirely, OR you can try this pattern below:

Something to try instead is more like:
```
try: 
   ...
except Exception as e:
    raise CustomErrorClass("message")
```

You might also like this package: https://docs.python.org/3/library/traceback.html#traceback-examples

File: preprocess.py [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+        lines = lines[columns]
+
+        lines.dropna(subset=['asin'], inplace=True)
+        lines.dropna(subset=['title'], inplace=True)
+        lines.dropna(subset=['category'], inplace=True)
+
+        lines.drop_duplicates(subset=['asin'], keep='first', inplace=True)
+        lines = lines[lines['title'].str.contains('\n')==False]
+
+        product_df = pd.concat([product_df, lines])
```
Comment (TensorDuck) [inline preprocess.py:76]: I think this block can use extra comments to explain why you're doing this cleaning. In general, docstrings would help too.

Also, implementing your own chunking is fine, but there are packages for handling dataframes that can't fit into memory: https://dask.org/

Might be too much for this project, so keep this for now. But something to look into for the future.

File: preprocess.py [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+
+    columns = ['asin', 'title', 'category']
+    product_df = pd.DataFrame(columns=columns)
+
+    for lines in chunk_df:
+        lines = lines[columns]
+
+        lines.dropna(subset=['asin'], inplace=True)
+        lines.dropna(subset=['title'], inplace=True)
+        lines.dropna(subset=['category'], inplace=True)
```
Comment (TensorDuck) [inline preprocess.py]: would this work: 
`lines.dropna(subset=["asin", "title", "category"], inplace=True)`

File: preprocess.py [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+    product_file = './data/source_data/meta_Books.json'
+    product_pkl = './data/meta_Book_product_df.pkl'
+
+    print('Start preprocess_review...')
+    preprocess_review(review_file, review_pkl)
+    print('Done preprocess_review\n')
+
+    print('Start preprocess_product...')
+    preprocess_product(product_file, product_pkl)
+    print('Done preprocess_review')
```
Comment (TensorDuck) [inline preprocess.py]: I'm wondering if this module does enough. It looks like you're mostly loading a DF, dropping some columns, and then converting to pickle.

**Comment by TensorDuck** [comment]
General Comments
* You need to add docstrings to the functions. Any number of formats would do, but it's hard to understand the purpose of a function just by reading the raw code. I'm quite partial to Google's docstring format: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html
* I don't see any object-oriented code. This would be a great challenge to making the code modular. I would recommend finding a way to write a class for your preprocessor. It might not do a lot, but it'll be more in line with the more common programming paradigms ML Engineers would be used to.

File: preprocess.py (hunk 12 above)
Comment (TensorDuck) [reply preprocess.py]: One mental model for this module:
The purpose of this module is to guarantee that any training or inference modules downstream have a consistent object to work with. 
To make the module easy to re-use, you should combine as many like-elements as possible. I think Both preprocess steps follow a constant process of:
1. load
2. Select columns
3. Drop NA
4. Other stuff
5. Save as pickle.

You can probably write a parent class that does 1-3 and 5, and then have child classes that handle 4.

File: preprocess.py (hunk 11 above)
Comment (TensorDuck) [reply preprocess.py:66]: A better practice would be selecting for the columns you want explicitly. Imagine this scenario:

Someone comes along and has their own data, but it has an extra column named "row_id". This would break downstream because this code doesn't remove an extra column you didn't expect.

File: preprocess.py [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+    try:
+        chunk_df = pd.read_json(product_file, lines=True,
+                                dtype={'asin': object}, chunksize=chunksize)
+    except:
+        print(f'Error loading {product_file}\nExit!')
+        sys.exit(1)
+
+    print(f'  Done reading {product_file} in chunksize')
+
+    columns = ['asin', 'title', 'category']
```
Comment (TensorDuck) [inline preprocess.py]: This is good! You should aim to select for the desired columns explicitly everywhere. You should do the same in your previous preprocess product step.

File: preprocess.py (hunk 11 above)
Comment (TensorDuck) [reply preprocess.py]: In fact, if you explicitly select for these columns, this should be even simpler:
`lines.dropna(inplace=True)`

File: build_save_model.py (hunk 3 above)
Comment (jt2002) [reply build_save_model.py]: I renamed to duplicate function name and push the update to dev.  Thank you.
